src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs

```diff
@@ -35,59 +78,128 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+            catch
+            {
+                CompleteAdding();
             }
-            catch (Exception) { }
         }
 
-        // for testing
-        internal void WriteMessage(LogMessageEntry entry)
+        private void ProcessLogQueue()
         {
-            IConsole console = entry.LogAsError ? ErrorConsole : Console;
-            console.Write(entry.Message);
+            while (TryDequeue(out LogMessageEntry message))
+            {
+                WriteMessage(message);
+            }
         }
 
-        private void ProcessLogQueue()
+        public bool Enqueue(LogMessageEntry item)
         {
-            try
+            lock (_messageQueue)
             {
-                foreach (LogMessageEntry message in _messageQueue.GetConsumingEnumerable())
+                while (_messageQueue.Count >= MaxQueueLength && !_isAddingCompleted)
+                {
+                    if (FullMode == ConsoleLoggerQueueFullMode.DropWrite)
+                    {
+                        _messagesDropped++;
+                        return true;
+                    }
+
+                    Debug.Assert(FullMode == ConsoleLoggerQueueFullMode.Wait);
+                    Monitor.Wait(_messageQueue);
+                }
+
+                if (!_isAddingCompleted)
                 {
-                    WriteMessage(message);
+                    Debug.Assert(_messageQueue.Count < MaxQueueLength);
+                    bool startedEmpty = _messageQueue.Count == 0;
+                    if (_messagesDropped > 0)
+                    {
+                        _messageQueue.Enqueue(new LogMessageEntry(
+                            message: SR.Format(SR.WarningMessageOnDrop + Environment.NewLine, _messagesDropped),
+                            logAsError: true
+                        ));
+
+                        _messagesDropped = 0;
+                    }
+
+                    // if we just logged the dropped message warning this could push the queue size to
+                    // MaxLength + 1, that shouldn't be a problem. It won't grow any further until it is less than
+                    // MaxLength once again.
+                    _messageQueue.Enqueue(item);
+
+                    // if the queue started empty it could be at 1 or 2 now
+                    if (startedEmpty)
+                    {
+                        // pulse for wait in Dequeue
+                        Monitor.PulseAll(_messageQueue);
+                    }
+
+                    return true;
                 }
             }
-            catch
+
+            return false;
+        }
+
+        public bool TryDequeue(out LogMessageEntry item)
+        {
+            lock (_messageQueue)
             {
-                try
+                while (_messageQueue.Count == 0 && !_isAddingCompleted)
+                {
+                    Monitor.Wait(_messageQueue);
+                }
+
+                if (_messageQueue.Count > 0 && !_isAddingCompleted)
```

Comment:
So after the `Dispose` is called - we won't log any messages that are still in buffer?
@vlada-shubina as FYI

Reply:
To elaborate more
* `Dispose` calls `CompleteAdding` which tries to lock `_messageQueue` and then marks `_isAddingCompleted` with `true`
* `ProcessLogQueue` calls `TryDequeue` in a loop, per single item. For each call it needs to lock `_messageQueue` to be able to return item to be logged by `ProcessLogQueue`
* If `_isAddingCompleted` is `true` the `TryDequeue` will return false - even if `_messageQueue.Count` is nonempty

Those facts bring the following concerns:
* The `CompleteAdding` and `TryDequeue` races for same lock. Once `CompleteAdding` wins, nothing more is logged, even if items are still queued - this seems to be a bug.
* `TryDequeue` tries to acquire lock for each single item being logged - I'm surprised this ends up being faster then using low-lock `BlockingCollection`. If it really is - should the `BlockingCollection` be actualy revised?

Reply:
@JanKrivanek if you can provide a sample code to demonstrate the concerns, you may open a new issue and we can track looking at that there. Thanks for your feedback.

Reply:
I am seeing you already logged https://github.com/dotnet/runtime/issues/79812. We'll try to look at that at some point.

Reply:
Sample repro: https://github.com/JanKrivanek/ConsoleLogging

Showing that messages are being dropped in 7.0 version of logging.